### ISSUE TITLE
pbrt-v4 is substantially slower than pbrt-v3

### ISSUE THREAD

**Issue by jiangwei007** [mmp/pbrt-v4#79]
When I'm testing on pbrt-v4 used the scene(lte-orb-silver.pbrt) is substantially slower than pbrt-v3,
The pbrt-v4 cpu path integrator is cost: 47.5s on my notebook
The pbrt-v3 path integrator is cost :20.4s on my notebook
the scene descriptor is:
```
Film "rgb"
    "string filename" "lte-orb-silver.exr"
    "integer yresolution" [ 400 ]
    "integer xresolution" [ 400 ]
Scale -1 1 1
LookAt -0.3 0.5 -0.5
    0 0.075 0
    0 1 0
Camera "perspective"
    "float fov" [ 37 ]
Sampler "sobol"
    "integer pixelsamples" [ 64 ]
Integrator "path" "integer maxdepth" 8

WorldBegin

ReverseOrientation

AttributeBegin
    AreaLightSource "diffuse"
        "rgb L" [ 9.5 9.5 9.5 ]
    ReverseOrientation
    Shape "trianglemesh"
        "integer indices" [ 0 1 2 0 2 3 ]
        "point3 P" [ -1 3.204597 -0.999721 1 3.204597 -0.999721 1 3.251826 0.999721 
                     -1 3.251826 0.999721 ]
AttributeEnd
# Material in the inner part of the sphere
# Matierial on the outer part of the sphere
MakeNamedMaterial "InnerMat"
    "string type" [ "coateddiffuse" ]
    "float roughness" .1
    "rgb reflectance" [ 0.2 0.3 0.7 ]
MakeNamedMaterial "OuterMat"
    "string type" [ "conductor" ]
    "float roughness" [ 0.0025 ]
    "spectrum k" [ "metal-Ag-k" ]
    "spectrum eta" [ "metal-Ag-eta" ]

Include "geometry/geometry.pbrt"
```

**Comment by jiangwei007** [comment]
i doubt that pointer dispatch mechanism ("TaggedPointer") is slower than pbrt-v3 used native virtual function dispatch, but need to verify that.

**Comment by jiangwei007** [comment]
I did a simple test, created 15 TaggedPointer Handle types, and 15 derived classes for virtual function calls, and made 100 million loop calls. They took close to the time. In most cases, the performance of TaggedPointer is a little better...
so cause of performance slowdown should not be TaggedPointer

**Comment by jiangwei007** [comment]
then i had tested SimplePathIntegrator, and the time it consumes close to PathIntegrator, so the length of integrator should not be the main reason for the significant performance slowdown compared to PBRT-V3.
@mmp would you point out the direction of the problem? thanks very much.

**Comment by mmp** [comment]
The good news is that although it is slower, it is nearly as efficient. :-)

I see about a 1.7x slowdown on my system here. However, the MSE is improved. I rendered reference images with 8k spp and then rendered with a uniform random sampler and 64 spp to measure MSE. Thanks to better sampling algorithms, pbrt-v4 has 1.4x lower MSE with this scene with the same number of samples. (Or, pbrt-v3 would need to run for 1.4 as long to get the same error.) The reduction in noise is evident in the images.

```
% imgtool diff v3/random64.exr --reference v3/ref.exr
Images differ:
	v3/random64.exr v3/ref.exr
	avg = 0.6702912 / 0.6705046 (-0.031833373% delta), MSE = 0.0069277845
% imgtool diff v4/random64.exr --reference v4/ref.exr
Images differ:
	v4/random64.exr v4/ref.exr
	avg = 0.6232793 / 0.6231396 (0.022411298% delta), MSE = 0.004969138
```

I did some profiling runs next.  Here is the top 95% of where pbrt-v4 spends its time:
![image](https://user-images.githubusercontent.com/102567/98449677-419a4f00-20ea-11eb-81e5-e6c901e5e80f.png)

And here is the same for pbrt-v3:
![image](https://user-images.githubusercontent.com/102567/98449680-4fe86b00-20ea-11eb-98a1-81914d26da1c.png)

It's pretty much all the same in terms of percentages, so it's not that one thing has gotten way slower, and we can see that the dynamic dispatch doesn't introduce a ton of overhead. Yet it is still 1.7/1.4=1.2x less efficient..

There are numerous differences in the implementation that could cause some loss of performance:
* It does spectral rendering now, so there is some overhead from that and in converting RGB to spectrum samples during rendering
* That "coateddiffuse" material actually does a random walk through layered media sampling the volume rendering equation for each and every BRDF evaluation, sampling operation, and PDF computation
* The light sampling algorithms are much improved (but more computationally intensive)
* The "volpath" integrator is the default; switching to "path" gives about 6% speedup. This is likely because "volpath" has some additional logic to perform MIS over wavelengths as part of its support for chromatic media. This is wasted work in the absence of chromatic media and is likely a lot of that 6%

So, while I wish it hadn't gotten a bit slower, I think it's worth it for the added functionality. Thanks for reporting this, though--it was good to dig into!
